### PR TITLE
[[ Bug 19271 ]] Expand range of 'random()' to 53-bits

### PR DESCRIPTION
**Pull request by abrisan**
Defined constants for limits of integer representation as double and increased input limit of random function to 2^53 - 1

### CODE REVIEW

**Comment by abrisan** [comment]
Depends on PR https://github.com/livecode/livecode/pull/5470

**Comment by livecodepanos** [comment]
Hi @alexbrisan,

It looks like you have branched off from `develop-8.1`, but you have submitted this PR against `develop`, so it includes all the other changes that exist in `develop-8.1`. 

I am not entirely sure what should be the target for this PR. Since it is an enhancement, it should go into `develop`, but I see no reason why not putting this into `develop-8.1`.

@runrevmark @livecodeali any thoughts?

**Comment by livecodeali** [comment]
@livecodepanos Don't worry - this PR is meant for develop and depends on a PR currently targeted at develop-8.1, so it'll have to wait until everything is merged up anyway

**Comment by livecodeali** [comment]
@alexbrisan It looks like this will have to be rebased after all - it's conflicting with changes made in develop.

**Comment by runrevmark** [comment]
What's needed to get this PR ready to merge? The testrunner change is in develop-8.1, can we cherry pick it, or merge up then rebase this?

**Comment by livecodeali** [comment]
The test runner change is already merged up, so it just needs rebasing

**Comment by runrevmark** [comment]
@livecode-vulcan review ok f3b71f2

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok f3b71f2a27c673e97cc71a99632d22bf0214a39a

**Comment by livecode-vulcan** [comment]
😞 test failure f3b71f2a27c673e97cc71a99632d22bf0214a39a

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/268)
* try-community-armv6-android-api9: [failure](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/277)
* try-community-js-emscripten-sdk1.35: [failure](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/257)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/200)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/200)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/266)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/271)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/270)
* try-community-x86-linux-debian8: [failure](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/275)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/270)
* try-community-x86_64-linux-debian8: [failure](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/275)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/294)
* try-community-x86_64-win32: [pending](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/265)

**Comment by abrisan** [comment]
@livecode-vulcan retry f3b71f2a27c673e97cc71a99632d22bf0214a39a

File: docs/development/build-linux.md [hunk 1]
```diff
@@ -1,85 +0,0 @@
-# Compiling LiveCode for Linux
```
Comment (livecodepanos) [inline docs/development/build-linux.md]: Is the removal of that file (`docs/development/build-linux.md`) part of this PR, or was it accidentally committed?

**Comment by livecodepanos** [comment]
@livecode-vulcan review by @runrevmark ok eed252c

File: engine/src/typedefs.h [hunk 2]
```diff
@@ -19,39 +19,6 @@ along with LiveCode.  If not see <http://www.gnu.org/licenses/>.  */
 
 // IEEE floating-point limits
```
Comment (runrevmark) [inline engine/src/typedefs.h]: You can remove this comment too :)

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok eed252c3d83f77bd316498088190b648ed2464b3

**Comment by livecodepanos** [comment]
Oops I spoke too soon :)

@livecode-vulcan review by @runrevmark not ok eed252c

**Comment by livecode-vulcan** [comment]
💔 review by @runrevmark not ok eed252c3d83f77bd316498088190b648ed2464b3

**Comment by runrevmark** [comment]
@livecode-vulcan review ok 5c59889

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok 5c59889d8add4d8cc240e2f17ecbb555e48ccd63

File: tests/lcs/core/math/math.livecodescript [hunk 3]
```diff
@@ -918,3 +918,37 @@ repeat with i = 1 to 10
 end repeat
 
 end TestMathArrayWrapArray
+
+on _RandomMaxLimitTestThrows
+
+	local tN1, tN2
+
+	put 2^53 into tN1
```
Comment (runrevmark) [inline tests/lcs/core/math/math.livecodescript]: I think 2^53 should be fine as that is the maximum number we can represent - so you need to use the next representable double *after* 2^53 to trigger a throw.

File: tests/lcs/core/math/math.livecodescript [hunk 4]
```diff
@@ -918,3 +918,37 @@ repeat with i = 1 to 10
...
+	
+	local rN1
+
+	put rand(tN1) into rN1
+	
+end _RandomMaxLimitTestThrows
+
+
+
+on _RandomMaxLimitTestNoThrow
```
Comment (runrevmark) [inline tests/lcs/core/math/math.livecodescript]: This test seems to fail as well - but it is not clear why.

File: engine/src/exec-math.cpp [hunk 5]
```diff
@@ -528,7 +528,7 @@ void MCMathEvalRandom(MCExecContext& ctxt, real64_t p_in, real64_t& r_result)
 	p_in = floor(p_in + 0.5);
 	// MW-2007-07-03: [[ Bug 4506 ]] - Large integers result in negative numbers
 	// being generated.
-	if (p_in < 1.0 || p_in >= 4294967296.0)
+	if (p_in < 1.0 || p_in >= DBL_INT_MAX)
```
Comment (runrevmark) [inline engine/src/exec-math.cpp]: We should be fine with > DBL_INT_MAX (although this means that the test isn't failing for the reason I thought!).

**Comment by livecode-vulcan** [comment]
😎 test success 5c59889d8add4d8cc240e2f17ecbb555e48ccd63

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/298)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/304)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/284)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/227)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/227)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/293)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/298)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/296)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/302)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/296)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/302)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/321)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/292)

File: docs/notes/bugfix-19271.md [hunk 6]
```diff
@@ -0,0 +1 @@
+# Increased random function input limit to 2^53 - 1
```
Comment (runrevmark) [inline docs/notes/bugfix-19271.md]: Its been increased to 2^53 - hasn't it?

File: tests/lcs/core/math/math.livecodescript [hunk 7]
```diff
@@ -918,3 +918,45 @@ repeat with i = 1 to 10
...
 
 end TestMathArrayWrapArray
+
+on _RandomMaxLimitTestThrows
+
+	local tX, tN2
+
+	put 2^53 into tX
+	
+	repeat with i = 1 to 128
```
Comment (runrevmark) [inline tests/lcs/core/math/math.livecodescript]: I suggest putting a comment here - 'This loop computes the next representable double after tX'

**Comment by runrevmark** [comment]
@livecode-vulcan review ok d961443

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok d961443b9256939edb653c5c24527ee774772ef1

**Comment by livecode-vulcan** [comment]
😎 test success d961443b9256939edb653c5c24527ee774772ef1

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/301)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/307)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/287)
* try-community-universal-ios-iphoneos10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos10.3/builds/230)
* try-community-universal-ios-iphonesimulator10.3: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator10.3/builds/230)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/296)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/301)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/299)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/305)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/299)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/305)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/324)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/295)

**Pull request merged**